### PR TITLE
[FLINK-16245] Decoupling user classloader from context classloader. 

### PR DESCRIPTION
**Pull request by AHeise**
This is a follow-up PR of https://github.com/apache/flink/pull/11303 after https://github.com/apache/flink/pull/11611 disabled it because of performance regression in blink e2e test.

## What is the purpose of the change

Decoupling user class loader from context classloader.

Thus, user classloader can be unloaded even though a reference on the context classloader outlives the user code.

## Brief change log

- Adding SafetyNetWrapperClassLoader
- Use it for FlinkUserCodeClassLoaders
- Make sure LocalExecutor closes user classloader

Compared to the original PR, it includes the following changes
- Use the parent classloader of the FlinkUserCodeClassLoader in SafetyNetWrapperClassLoader to address the performance concerns.
- Further, improve performance by enabling parallel class loading in all involved classloaders.

## Verifying this change

Added unit test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zentol** [comment]
Could you summarize what the problem was that caused the performance regression?

**Comment by AHeise** [comment]
> Could you summarize what the problem was that caused the performance regression?

For some reason, not setting a parent classloader in the wrapper is much slower. I cannot see an obvious reason and added a couple of measurements.

First, I wanted to know if the performance difference was caused by slower class loading. So I measured the total time spent in `ClassLoader#loadClass` for the previous version and this version. I found no difference. There is also no difference by the amount of classes loaded.

Second, a profiler revealed that much time in the slow version was spent in `Class#newInstance`. I added the time measurements to `org.apache.flink.table.runtime.generated.GeneratedClass`. Interestingly, there is again no difference in compiling the class. There is a huge difference in calling `newInstance` on the generated class though that can easily account for the time difference. 

Without the parent, creating an instance of the 4 specific blink operators `BatchNestedLoopJoin`, `LongHashJoinOperator`, `LocalHashAggregateWithKeys`, and `HashAggregateWithKeys` slows down over time taking up over 10s to finish at the end of the. 

Interestingly, they use only the instantiation method with explicit arguments `GeneratedClass#newInstance(ClassLoader classLoader, Object... args)`. However, there are also operators that use that method that are "well-behaved".

I have also forced class resolution through reflection on the generated classes, but resolution is fast in all cases. It's really just about creating the instances.

**Comment by AHeise** [comment]
There was also a [classloader leak in ORC](https://issues.apache.org/jira/browse/ORC-653), for which I added a workaround.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java [hunk 1]
```diff
@@ -241,12 +242,14 @@ public final DataStreamSink consumeDataStream(DataStream dataStream) {
...
 		RowType formatType = RowType.of(formatTypes, formatNames);
-		Configuration formatConf = new Configuration(jobConf);
-		sd.getSerdeInfo().getParameters().forEach(formatConf::set);
 		if (serLib.contains("parquet")) {
+			Configuration formatConf = new Configuration(jobConf);
+			sd.getSerdeInfo().getParameters().forEach(formatConf::set);
 			return Optional.of(ParquetRowDataBuilder.createWriterFactory(
 					formatType, formatConf, hiveVersion.startsWith("3.")));
 		} else if (serLib.contains("orc")) {
+			Configuration formatConf = new ThreadLocalClassLoaderConfiguration(jobConf);
```
Comment (zentol) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java:251]: If we were to use a normal configuration, what would happen? Subsequent jobs would not be able to run?

Comment (AHeise) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java:251]: Yes. That makes me believe that we should probably make the safety net configurable (on by default). Or would you think we can catch up and fix leaks much quicker?

Comment (zentol) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java:251]: An opt-out switch is a good idea; also for cases where things don't behave as expected close to the next release, and we don't want to revert commits at that point.

File: flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java [hunk 2]
```diff
@@ -124,6 +124,15 @@
 		return parseParentFirstLoaderPatterns(base, append);
 	}
 
+	@Documentation.Section(Documentation.Sections.EXPERT_CLASS_LOADING)
+	public static final ConfigOption<Boolean> CHECK_LEAKED_CLASSLOADER = ConfigOptions
+		.key("classloader.check_leaked_classloader")
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java:129]: we don't use underscores.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java [hunk 3]
```diff
@@ -136,7 +150,10 @@ public void close() {
 
 		private FlinkUserCodeClassLoader ensureInner() {
 			if (inner == null) {
-				throw new IllegalStateException("Trying to access closed classloader");
+				throw new IllegalStateException("Trying to access closed classloader. Please check if you store " +
+					"classloaders directly or indirectly in static fields. If the stacktrace suggests that the leak " +
+					"occurs in a third party library and cannot be fixed immediately, you can disable this check " +
+					"with the configuration 'classloader.check_leaked_classloader'.");
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java:156]: Retrieve the key from the actual configuration object instead.

File: flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java [hunk 4]
```diff
@@ -124,6 +124,15 @@
...
+		.booleanType()
+		.defaultValue(true)
+		.withDescription("Fails if a user classloader is used in another job after a job has been finished. This " +
+			"check should only be disabled if the class leak occurs in an external library cannot be fixed " +
+			"immediately.");
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java:134]:
```
Fails attempts at loading classes if the user classloader of a job is used after it has terminated.
This is usually caused by the classloader being leaked by lingering threads or misbehaving libraries, which may also result in the classloader being used by other jobs.
This check should only be disabled if such a leak prevents further jobs from running.
```
